File: Src/Photonne.Client.Web/Models/MapCluster.cs
using Photonne.Client.Web;

namespace Photonne.Client.Web.Models;

public class MapCluster
{
    public string Id { get; set; } = string.Empty;
    public double Latitude { get; set; }
    public double Longitude { get; set; }
    public int Count { get; set; }
    public List<Guid> AssetIds { get; set; } = new();
    public DateTime EarliestDate { get; set; }
    public DateTime LatestDate { get; set; }
    public Guid? FirstAssetId { get; set; }
    public bool HasThumbnail { get; set; }
    
    public string ThumbnailUrl => FirstAssetId.HasValue && HasThumbnail
        ? $"{ApiConfig.BaseUrl}/api/assets/{FirstAssetId.Value}/thumbnail?size=Medium"
        : string.Empty;

    public override string ToString()
    {
        return $"Cluster: {Latitude:F4}, {Longitude:F4}, Count: {Count}, Date: {EarliestDate:yyyy-MM-dd}";
    }
}

File: Src/Photonne.Client.Native/composeApp/src/commonMain/kotlin/com/photonne/app/data/models/MapCluster.kt
package com.photonne.app.data.models

import kotlinx.datetime.Instant
import kotlinx.serialization.Serializable

/**
 * A pre-clustered group of GPS-tagged assets. Mirrors the backend's
 * `MapClusterResponse`: positions are pre-aggregated and visually
 * de-overlapped server-side based on the requested zoom level, so the
 * client only renders.
 */
@Serializable
data class MapCluster(
    val id: String,
    val latitude: Double,
    val longitude: Double,
    val count: Int = 1,
    val assetIds: List<String> = emptyList(),
    @Serializable(with = FlexibleInstantSerializer::class)
    val earliestDate: Instant,
    @Serializable(with = FlexibleInstantSerializer::class)
    val latestDate: Instant,
    val firstAssetId: String? = null,
    val hasThumbnail: Boolean = false
)
